### PR TITLE
[FIX] web_editor: emoji picker popover mispositioned

### PR DESCRIPTION
**Pull request by dhba-odoo**
**Current behavior before PR:**

- Emoji picker opened at an incorrect position.

**Desired behavior after PR is merged:**

- The Emoji picker now opens precisely at the cursor location

task-3569967

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 1]
```diff
@@ -1533,25 +1533,30 @@ export class Wysiwyg extends Component {
...
 
         this.popover.add(closest, EmojiPicker, {
                 onSelect: (str) => {
                     restoreSelection();
                     this.odooEditor.execCommand('insert', str);
                 },
+            }, {
+                onPositioned: (popover, position) => {
+                    restoreSelection();
+                    const rangePosition = getRangePosition(popover, this.options.document, this.options.getContextFromParentRect());
```
Comment (cammarosano) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1545]: Hello! The position of the emoji picker seems broken in mass_mailing.

I suspect it's because of a problem with `options.document` in mass_mailing's Wysiwyg, that gets (wrongly) assigned to the top document instead of the iframe's document after the Wysiwyg component is re-rendered.
Considering that https://github.com/odoo/odoo/pull/138259, which aims to fix it, was recently merged into saas-16.4, you might wanna give it a little test after rebasing your branch.

Alternatively, you could use `this.odooEditor.document` instead of `this.options.document` here.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 2]
```diff
@@ -1533,25 +1533,26 @@ export class Wysiwyg extends Component {
...
                     restoreSelection();
                     this.odooEditor.execCommand('insert', str);
                 },
+            }, {
+                onPositioned: (popover) => {
+                    restoreSelection();
+                    let options = {};
+                    if (this.options.getContextFromParentRect) {
+                        options['parentContextRect'] = this.options.getContextFromParentRect();
+                    }
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1548]: IMO this piece serves no purpose here, since we are not using it here,  also there is room for this code to be optimized.
i suggest this code will also do the job :
`                onPositioned: (popover) => {
                    restoreSelection();
                    const rangePosition = getRangePosition(popover, this.options.document, {});
                    popover.style.top = rangePosition.top + 'px';
                    popover.style.left = rangePosition.left + 'px';
                    const oInputBox = popover.getElementsByClassName('o_input')[0];
                    oInputBox.focus();
                },`
@dhba-odoo can you check it let me know.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1548]: @shsa-odoo I actually require the code as it is. In the 'getDeepRange' method, there is an 'if' condition that is essential for handling the 'parentContextRect.' If this condition is removed, the code within 'getDeepRange' will not execute, leading to an incorrect positioning of the emoji picker in Email Marketing. Therefore, we cannot remove that code.

Comment (shsa-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1548]: I tried functionally by removing that code and it seems to work fine in editor. What am i missing ? is it missing something in email marketing ?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 3]
```diff
@@ -1563,25 +1563,28 @@ export class Wysiwyg extends Component {
...
         this.popover.add(closest, EmojiPicker, {
                 onSelect: (str) => {
                     restoreSelection();
                     this.odooEditor.execCommand('insert', str);
                 },
+            }, {
+                onPositioned: (popover) => {
+                    restoreSelection();
+                    // Sets the 'parentContextRect' option in 'options' when 'getContextFromParentRect' is available.
+                    // This facilitates element positioning relative to a parent or reference rectangle.
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1576]: Character limit.

```suggestion
                    // Set the 'parentContextRect' option in 'options' when
                    // 'getContextFromParentRect' is available. This facilitates
                    // element positioning relative to a parent or reference
                    // rectangle.
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 4]
```diff
@@ -1563,25 +1563,28 @@ export class Wysiwyg extends Component {
...
+                    let options = {};
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1577]: This is a constant.

```suggestion
                    const options = {};
```

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 5]
```diff
@@ -1563,25 +1563,28 @@ export class Wysiwyg extends Component {
...
-        // Calculate one/fouth portion of blockWidth, and then on the basis of our cursor position
-        // check in which quadrant it falls, give the popover its position accordingly
-        const portionWidth = (blockWidth / 4);
-        const cursorAt = getRangePosition(targetEl, this.odooEditor.document).left;
-        let position = 'end';
-        if (cursorAt < portionWidth * 2) {
-            position = 'start';
-        } else if (cursorAt <= portionWidth * 3) {
-            position = 'middle';
-        }
```
Comment (Zynton) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1576]: This seems like it would be useful. Why did you remove it?

**Review by Zynton — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1576]: It's used to determine the popover's position based on the cursor position, but it has limitations, as it only supports three fixed positions (start, middle, end) of block. 
and that is why I have used `onPositioned` in which the popover positioned at cursor position.
So, we no longer require this as the popover position will be determined within the `onPositioned`.

Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1587]: But didn't this ensure the popover couldn't get out of bounds of the viewport?

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1587]: Yes, this code ensures the popover stays within the `viewport`. However, even if I don't remove this code, the popover position set by it will be overwritten by my changes.

Additionally, I haven't encountered any situations where the popover goes out of bounds. I'm using the same positioning function as the `powerbox`, and since the `powerbox` doesn't have any issues like this, I believe the popover should be fine too.

Comment (Zynton) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1587]: I tested and had no issues indeed. Thanks for the clarifications :+1:

**Comment by Zynton** [comment]
LGTM

@robodoo r+
